File: app/src/main/java/com/example/a4/MainActivity.java
package com.example.a4;

import android.animation.Animator;
import android.app.AlertDialog;
import android.content.ContentValues;
import android.content.Context;
import android.content.DialogInterface;
import android.content.Intent;
import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.graphics.drawable.Drawable;
import android.os.Build;
import android.os.Bundle;

import com.google.android.material.floatingactionbutton.FloatingActionButton;
import com.google.android.material.snackbar.Snackbar;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.Toolbar;
import androidx.constraintlayout.widget.ConstraintLayout;
import androidx.coordinatorlayout.widget.CoordinatorLayout;
import androidx.core.content.ContextCompat;
import androidx.core.graphics.drawable.DrawableCompat;
import androidx.recyclerview.widget.DividerItemDecoration;
import androidx.recyclerview.widget.ItemTouchHelper;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import android.util.DisplayMetrics;
import android.util.Log;
import android.view.MenuInflater;
import android.view.View;
import android.view.Menu;
import android.view.MenuItem;
import android.view.ViewAnimationUtils;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.ImageView;
import android.widget.SearchView;
import android.widget.Spinner;
import android.widget.Toast;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class MainActivity extends AppCompatActivity implements SearchView.OnCloseListener {
    RecyclerView recyclerView;
    Adaptor adaptor;
    DataClass dataClass;
    DbHelper dbHelper;
    String sortMode= dbHelper.COLUMN_ID + " DESC" ;
    SearchView searchView;
    Toolbar toolbar;
    List<DataClass> notesList;
    CoordinatorLayout coordinatorLayout;
    public static final float ALPHA_FULL = 1.0f;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main3);
        hideHomeButton();
        init();
        setSupportActionBar(toolbar);
    }


    private void list() {
        adaptor = new Adaptor(this);
        recyclerView.setLayoutManager(new LinearLayoutManager(this));
        recyclerView.setAdapter(adaptor);
        ItemTouchHelper helper = new ItemTouchHelper(new ItemTouchHelper.SimpleCallback(0, ItemTouchHelper.RIGHT | ItemTouchHelper.LEFT) {
            @Override
            public boolean onMove(@NonNull RecyclerView recyclerView,
                                  @NonNull RecyclerView.ViewHolder viewHolder, @NonNull RecyclerView.ViewHolder target) {
                return false;
            }

            public void onChildDraw(Canvas c, RecyclerView recyclerView, RecyclerView.ViewHolder viewHolder, float dX, float dY, int actionState, boolean isCurrentlyActive) {
                if (actionState == ItemTouchHelper.ACTION_STATE_SWIPE) {
                    // Get RecyclerView item from the ViewHolder
                    View itemView = viewHolder.itemView;

                    Paint p = new Paint();
                    Bitmap icon;

                    if (dX > 0) {
            /* Note, ApplicationManager is a helper class I created
               myself to get a context outside an Activity class -
               feel free to use your own method */

                        icon = getBitmapFromVectorDrawable(MainActivity.this,R.drawable.ic_edit);

                        /* Set your color for positive displacement */
                        p.setARGB(255, 0, 255, 0);

                        // Draw Rect with varying right side, equal to displacement dX
                        c.drawRect((float) itemView.getLeft(), (float) itemView.getTop(), dX,
                                (float) itemView.getBottom(), p);

                        // Set the image icon for Right swipe
                        c.drawBitmap(icon,
                                (float) itemView.getLeft() + convertDpToPx(16),
                                (float) itemView.getTop() + ((float) itemView.getBottom() - (float) itemView.getTop() - icon.getHeight()) / 2,
                                p);
                    } else {
                        icon = getBitmapFromVectorDrawable(MainActivity.this,R.drawable.ic_delete);

                        /* Set your color for negative displacement */
                        p.setARGB(255, 255, 0, 0);

                        // Draw Rect with varying left side, equal to the item's right side
                        // plus negative displacement dX
                        c.drawRect((float) itemView.getRight() + dX, (float) itemView.getTop(),
                                (float) itemView.getRight(), (float) itemView.getBottom(), p);

                        //Set the image icon for Left swipe
                        c.drawBitmap(icon,
                                (float) itemView.getRight() - convertDpToPx(16) - icon.getWidth(),
                                (float) itemView.getTop() + ((float) itemView.getBottom() - (float) itemView.getTop() - icon.getHeight()) / 2,
                                p);
                    }

                    // Fade out the view as it is swiped out of the parent's bounds
                    final float alpha = ALPHA_FULL - Math.abs(dX) / (float) viewHolder.itemView.getWidth();
                    viewHolder.itemView.setAlpha(alpha);
                    viewHolder.itemView.setTranslationX(dX);

                } else {
                    super.onChildDraw(c, recyclerView, viewHolder, dX, dY, actionState, isCurrentlyActive);
                }
            }

            private int convertDpToPx(int dp) {
                return Math.round(dp * (getResources().getDisplayMetrics().xdpi / DisplayMetrics.DENSITY_DEFAULT));
            }


            @Override
            public void onSwiped(@NonNull RecyclerView.ViewHolder viewHolder, int direction) {
             final int position = viewHolder.getAdapterPosition();
                final DataClass dataClass4 = notesList.get(position);
                if (direction == ItemTouchHelper.LEFT) {
                    Thread thread4 = new Thread(new Runnable() {
                        @Override
                        public void run() {
                            dbHelper.deleteData(String.valueOf(dataClass4.getId()));
                            readData();
                            Snackbar snackbar = Snackbar
                                    .make(coordinatorLayout, "Item was removed from the list.", Snackbar.LENGTH_LONG);
                            snackbar.setAction("UNDO", new View.OnClickListener() {
                                @Override
                                public void onClick(View view) {
                                    Thread thread4 = new Thread(new Runnable() {
                                        @Override
                                        public void run() {
                                            SQLiteDatabase db = dbHelper.getWritableDatabase();
                                            ContentValues contentValues = new ContentValues();
                                            contentValues.put(DbHelper.COLUMN_NAME_TITLE,dataClass4.getNoteName());
                                            contentValues.put(DbHelper.COLUMN_NAME_NOTE,dataClass4.getNote());
                                            contentValues.put(DbHelper.COLUMN_DATE,dataClass4.getDate());
                                            contentValues.put(DbHelper.COLUMN_ID,dataClass4.getId());
                                            long id =db.insert(DbHelper.TABLE_NAME,null,contentValues);
                                            Log.d("tag",String.valueOf(id));
                                            db.close();
                                            readData();
                                        }
                                    });
                                    thread4.start();
                                    recyclerView.scrollToPosition(position);
                                }
                            });
                            snackbar.setActionTextColor(Color.YELLOW);
                            snackbar.show();
                        }
                    });
                    thread4.start();
                } else {
                    Intent intent = new Intent(MainActivity.this,EditActivity.class);
                    intent.putExtra("note",dataClass4.getNote());
                    intent.putExtra("noteName",dataClass4.getNoteName());
                    intent.putExtra("id",dataClass4.getId());
                    startActivity(intent);
                }
            }
        });
        helper.attachToRecyclerView(recyclerView);
    }

    private void init() {
        toolbar = findViewById(R.id.toolbar);
        recyclerView = findViewById(R.id.recyclerView);
        final FloatingActionButton fab = findViewById(R.id.fab);
        fab.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                startActivity(new Intent(MainActivity.this, WriteActivity.class));
            }
        });
        dataClass = new DataClass();
        dbHelper = new DbHelper(this);
        coordinatorLayout = findViewById(R.id.coordi);
    }

    public void readData() {
        notesList = new ArrayList<>();
        Thread thread2 = new Thread(new Runnable() {
            @Override
            public void run() {
                SQLiteDatabase db = dbHelper.getReadableDatabase();
                Cursor cursor = db.query(dbHelper.TABLE_NAME, null, null, null,
                        null, null, sortMode);
                while (cursor.moveToNext()) {
                    String noteName = cursor.getString(
                            cursor.getColumnIndexOrThrow(DbHelper.COLUMN_NAME_TITLE));
                    String note = cursor.getString(
                            cursor.getColumnIndexOrThrow(DbHelper.COLUMN_NAME_NOTE));
                    String date = cursor.getString(
                            cursor.getColumnIndexOrThrow(DbHelper.COLUMN_DATE));
                    Long id = cursor.getLong(cursor.getColumnIndexOrThrow(dbHelper.COLUMN_ID));
                    DataClass dataClass = new DataClass();
                    dataClass.setNoteName(noteName);
                    dataClass.setNote(note);
                    dataClass.setId(id);
                    dataClass.setDate(date);
                    Log.d("tag5", dataClass.getNote());
                    notesList.add(dataClass);
                }
                MainActivity.this.runOnUiThread(new Runnable() {
                    public void run() {
                        adaptor.submitList(notesList);
                    }
                });
                cursor.close();
                db.close();
            }
        });
        thread2.start();
    }

    @Override
    protected void onResume() {
        super.onResume();
        readData();
        list();
    }

    void searchData() {
        searchView.setOnQueryTextListener(new SearchView.OnQueryTextListener() {
            @Override
            public boolean onQueryTextSubmit(String query) {
                return false;
            }

            @Override
            public boolean onQueryTextChange(final String newText) {
                if (newText != null) {
                    Thread thread3 = new Thread(new Runnable() {
                        @Override
                        public void run() {
                            final List<DataClass> notesList = new ArrayList<>();
                            SQLiteDatabase db = dbHelper.getReadableDatabase();
                            String searchText = "%" + newText + "%";
                            Cursor cursor = db.query(dbHelper.TABLE_NAME, null,
                                    dbHelper.COLUMN_NAME_TITLE + " LIKE ? OR " +
                                            dbHelper.COLUMN_NAME_NOTE + " LIKE ?", new String[]{searchText, searchText},
                                    null, null, sortMode);
                            while (cursor.moveToNext()) {
                                String noteName = cursor.getString(
                                        cursor.getColumnIndexOrThrow(DbHelper.COLUMN_NAME_TITLE));
                                String note = cursor.getString(
                                        cursor.getColumnIndexOrThrow(DbHelper.COLUMN_NAME_NOTE));
                                String date = cursor.getString(
                                        cursor.getColumnIndexOrThrow(DbHelper.COLUMN_DATE));
                                Long id = cursor.getLong(cursor.getColumnIndexOrThrow(dbHelper.COLUMN_ID));
                                DataClass dataClass2 = new DataClass();
                                dataClass2.setNoteName(noteName);
                                dataClass2.setDate(date);
                                dataClass2.setNote(note);
                                dataClass2.setId(id);
                                Log.d("tag5", dataClass.getNote());
                                notesList.add(dataClass2);
                            }
                            MainActivity.this.runOnUiThread(new Runnable() {
                                public void run() {
                                    adaptor.submitList(notesList);
                                }
                            });
                            cursor.close();
                            db.close();
                        }
                    });
                    thread3.start();
                }
                return false;
            }
        });
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        MenuInflater inflater = getMenuInflater();
        inflater.inflate(R.menu.menu_main, menu);
        MenuItem sortItem = menu.findItem(R.id.action_sort);
        sortItem.setOnMenuItemClickListener(new MenuItem.OnMenuItemClickListener() {
            @Override
            public boolean onMenuItemClick(MenuItem item) {
                String[] sortList = getResources().getStringArray(R.array.sort_array);
                new AlertDialog.Builder(MainActivity.this).setTitle("sort by")
                        .setSingleChoiceItems(sortList,-1, new DialogInterface.OnClickListener() {
                    @Override
                    public void onClick(DialogInterface dialog, int which) {
                        if (which == 0) sortMode = dbHelper.COLUMN_ID + " ASC";
                        else if (which== 1) sortMode = dbHelper.COLUMN_ID + " DESC";
                        else sortMode = dbHelper.COLUMN_NAME_TITLE + " ASC";
                    }
                }).setPositiveButton("ok", new DialogInterface.OnClickListener() {
                    @Override
                    public void onClick(DialogInterface dialog, int which) {
                        readData();
                    }
                }).show();
                return false;
            }
        });
        MenuItem searchItem = menu.findItem(R.id.action_search);
        searchView = (SearchView) searchItem.getActionView();
        searchView.setOnCloseListener(this);
        searchData();
        return true;
    }

    @Override
    public boolean onClose() {
        readData();
        return false;
    }

    public static Bitmap getBitmapFromVectorDrawable(Context context, int drawableId) {
        Drawable drawable = ContextCompat.getDrawable(context, drawableId);
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.LOLLIPOP) {
            drawable = (DrawableCompat.wrap(drawable)).mutate();
        }

        Bitmap bitmap = Bitmap.createBitmap(drawable.getIntrinsicWidth(),
                drawable.getIntrinsicHeight(), Bitmap.Config.ARGB_8888);
        Canvas canvas = new Canvas(bitmap);
        drawable.setBounds(0, 0, canvas.getWidth(), canvas.getHeight());
        drawable.draw(canvas);

        return bitmap;
    }

    public void hideHomeButton(){
        if (getSupportActionBar()!=null) {
            MainActivity.this.getSupportActionBar().setHomeButtonEnabled(false); // disable the button
            MainActivity.this.getSupportActionBar().setDisplayHomeAsUpEnabled(false); // remove the left caret
            MainActivity.this.getSupportActionBar().setDisplayShowHomeEnabled(false);
        }
    }
}

File: app/src/main/java/com/example/a4/MainActivity.kt
package com.example.a4

import android.app.AlertDialog
import android.content.ContentValues
import android.content.Context
import android.content.Intent
import android.graphics.Bitmap
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.os.Build
import android.os.Bundle
import android.util.DisplayMetrics
import android.util.Log
import android.view.Menu
import android.widget.SearchView
import androidx.appcompat.app.AppCompatActivity
import androidx.core.content.ContextCompat
import androidx.core.graphics.drawable.DrawableCompat
import androidx.recyclerview.widget.ItemTouchHelper
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.google.android.material.snackbar.Snackbar
import kotlinx.android.synthetic.main.activity_main3.*
import java.util.*

class MainActivity : AppCompatActivity(), SearchView.OnCloseListener {
    lateinit var adaptor: Adaptor
    lateinit var notesDataClass: Notes
    lateinit var dbHelper: DbHelper
    var sortMode = DbHelper.COLUMN_ID + " DESC"
    var searchView: SearchView? = null
    var notesList: ArrayList<Notes>? = null
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main3)
        hideHomeButton()
        init()
        setSupportActionBar(toolbarMain)
    }

    private fun list() {
        adaptor = Adaptor(this)
        recyclerView.layoutManager = LinearLayoutManager(this)
        recyclerView.adapter = adaptor
        val helper = ItemTouchHelper(object : ItemTouchHelper.SimpleCallback(0, ItemTouchHelper.RIGHT or ItemTouchHelper.LEFT) {
            override fun onMove(recyclerView: RecyclerView,
                                viewHolder: RecyclerView.ViewHolder, target: RecyclerView.ViewHolder): Boolean {
                return false
            }

            override fun onChildDraw(c: Canvas, recyclerView: RecyclerView, viewHolder: RecyclerView.ViewHolder, dX: Float, dY: Float, actionState: Int, isCurrentlyActive: Boolean) {
                if (actionState == ItemTouchHelper.ACTION_STATE_SWIPE) {
                    // Get RecyclerView item from the ViewHolder
                    val itemView = viewHolder.itemView
                    val p = Paint()
                    val icon: Bitmap
                    if (dX > 0) {
                        /* Note, ApplicationManager is a helper class I created
               myself to get a context outside an Activity class -
               feel free to use your own method */
                        icon = getBitmapFromVectorDrawable(this@MainActivity, R.drawable.ic_edit)

                        /* Set your color for positive displacement */p.setARGB(255, 0, 255, 0)

                        // Draw Rect with varying right side, equal to displacement dX
                        c.drawRect(itemView.left.toFloat(), itemView.top.toFloat(), dX,
                                itemView.bottom.toFloat(), p)

                        // Set the image icon for Right swipe
                        c.drawBitmap(icon,
                                itemView.left.toFloat() + convertDpToPx(16),
                                itemView.top.toFloat() + (itemView.bottom.toFloat() - itemView.top.toFloat() - icon.height) / 2,
                                p)
                    } else {
                        icon = getBitmapFromVectorDrawable(this@MainActivity, R.drawable.ic_delete)

                        /* Set your color for negative displacement */p.setARGB(255, 255, 0, 0)

                        // Draw Rect with varying left side, equal to the item's right side
                        // plus negative displacement dX
                        c.drawRect(itemView.right.toFloat() + dX, itemView.top.toFloat(),
                                itemView.right.toFloat(), itemView.bottom.toFloat(), p)

                        //Set the image icon for Left swipe
                        c.drawBitmap(icon,
                                itemView.right.toFloat() - convertDpToPx(16) - icon.width,
                                itemView.top.toFloat() + (itemView.bottom.toFloat() - itemView.top.toFloat() - icon.height) / 2,
                                p)
                    }

                    // Fade out the view as it is swiped out of the parent's bounds
                    val alpha = ALPHA_FULL - Math.abs(dX) / viewHolder.itemView.width.toFloat()
                    viewHolder.itemView.alpha = alpha
                    viewHolder.itemView.translationX = dX
                } else {
                    super.onChildDraw(c, recyclerView, viewHolder, dX, dY, actionState, isCurrentlyActive)
                }
            }

            private fun convertDpToPx(dp: Int): Int {
                return Math.round(dp * (resources.displayMetrics.xdpi / DisplayMetrics.DENSITY_DEFAULT))
            }

            override fun onSwiped(viewHolder: RecyclerView.ViewHolder, direction: Int) {
                val position = viewHolder.adapterPosition
                val dataClass4 = notesList!![position]
                if (direction == ItemTouchHelper.LEFT) {
                    val thread4 = Thread(Runnable {
                        dbHelper.deleteData(dataClass4.id.toString())
                        readData()
                        val snackbar = Snackbar
                                .make(coordi, "Item was removed from the list.", Snackbar.LENGTH_LONG)
                        snackbar.setAction("UNDO") {
                            val thread4 = Thread(Runnable {
                                val db = dbHelper.writableDatabase
                                val contentValues = ContentValues().apply {
                                    put(DbHelper.COLUMN_NAME_TITLE, dataClass4.noteName)
                                    put(DbHelper.COLUMN_NAME_NOTE, dataClass4.note)
                                    put(DbHelper.COLUMN_IMAGE_URI, dataClass4.imageUri)
                                    put(DbHelper.COLUMN_DATE, dataClass4.date)
                                    put(DbHelper.COLUMN_ID, dataClass4.id)
                                }
                                val id = db.insert(DbHelper.TABLE_NAME, null, contentValues)
                                Log.d("tag", id.toString())
                                db.close()
                                readData()
                            })
                            thread4.start()
                            recyclerView.scrollToPosition(position)
                        }
                        snackbar.setActionTextColor(Color.YELLOW)
                        snackbar.show()
                    })
                    thread4.start()
                } else {
                    val intent = Intent(this@MainActivity, EditActivity::class.java).apply {
                        putExtra("note", dataClass4.note)
                        if (dataClass4.imageUri != null) putExtra("uri", dataClass4.imageUri)
                        putExtra("noteName", dataClass4.noteName)
                        putExtra("id", dataClass4.id)
                    }
                    startActivity(intent)
                }
            }
        })
        helper.attachToRecyclerView(recyclerView)
    }

    private fun init() {
        fab.setOnClickListener { startActivity(Intent(this@MainActivity, WriteActivity::class.java)) }
        notesDataClass = Notes()
        dbHelper = DbHelper(this)
    }

    fun readData() {
        notesList = ArrayList()
        val thread2 = Thread(Runnable {
            val db = dbHelper.readableDatabase
            val cursor = db.query(DbHelper.TABLE_NAME, null, null, null,
                    null, null, sortMode)
            while (cursor.moveToNext()) {
                val noteNameString = cursor.getString(
                        cursor.getColumnIndexOrThrow(DbHelper.COLUMN_NAME_TITLE))
                val noteString = cursor.getString(
                        cursor.getColumnIndexOrThrow(DbHelper.COLUMN_NAME_NOTE))
                val dateString = cursor.getString(
                        cursor.getColumnIndexOrThrow(DbHelper.COLUMN_DATE))
                val idLong = cursor.getLong(cursor.getColumnIndexOrThrow(DbHelper.COLUMN_ID))
                val uriString: String? = cursor.getString(cursor.getColumnIndexOrThrow(DbHelper.COLUMN_IMAGE_URI))
                val dataClass = Notes().apply {
                    noteName = noteNameString
                    note = noteString
                    id = idLong
                    date = dateString
                    imageUri = uriString
                }
                Log.d("tag5", dataClass.note)
                notesList!!.add(dataClass)
            }
            runOnUiThread { adaptor.submitList(notesList) }
            cursor.close()
            db.close()
        })
        thread2.start()
    }

    override fun onResume() {
        super.onResume()
        readData()
        list()
    }

    private fun searchData() {
        searchView!!.setOnQueryTextListener(object : SearchView.OnQueryTextListener {
            override fun onQueryTextSubmit(query: String): Boolean {
                return false
            }

            override fun onQueryTextChange(newText: String): Boolean {
                if (newText != null) {
                    val thread3 = Thread(Runnable {
                        val notesList: MutableList<Notes> = ArrayList()
                        val db = dbHelper.readableDatabase
                        val searchText = "%$newText%"
                        val cursor = db.query(DbHelper.TABLE_NAME, null,
                                DbHelper.COLUMN_NAME_TITLE + " LIKE ? OR " +
                                        DbHelper.COLUMN_NAME_NOTE + " LIKE ?", arrayOf(searchText, searchText),
                                null, null, sortMode)
                        while (cursor.moveToNext()) {
                            val noteName = cursor.getString(
                                    cursor.getColumnIndexOrThrow(DbHelper.COLUMN_NAME_TITLE))
                            val note = cursor.getString(
                                    cursor.getColumnIndexOrThrow(DbHelper.COLUMN_NAME_NOTE))
                            val date = cursor.getString(
                                    cursor.getColumnIndexOrThrow(DbHelper.COLUMN_DATE))
                            val id = cursor.getLong(cursor.getColumnIndexOrThrow(DbHelper.COLUMN_ID))
                            val dataClass2 = Notes()
                            dataClass2.noteName = noteName
                            dataClass2.date = date
                            dataClass2.note = note
                            dataClass2.id = id
                            Log.d("tag5", notesDataClass.note)
                            notesList.add(dataClass2)
                        }
                        runOnUiThread { adaptor.submitList(notesList) }
                        cursor.close()
                        db.close()
                    })
                    thread3.start()
                }
                return false
            }
        })
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        val inflater = menuInflater
        inflater.inflate(R.menu.menu_main, menu)
        val sortItem = menu.findItem(R.id.action_sort)
        sortItem.setOnMenuItemClickListener {
            val sortList = resources.getStringArray(R.array.sort_array)
            AlertDialog.Builder(this@MainActivity).setTitle("sort by")
                    .setSingleChoiceItems(sortList, -1) { dialog, which -> sortMode = if (which == 0) DbHelper.COLUMN_ID + " ASC" else if (which == 1) DbHelper.COLUMN_ID + " DESC" else DbHelper.COLUMN_NAME_TITLE + " ASC" }.setPositiveButton("ok") { dialog, which -> readData() }.show()
            false
        }
        val searchItem = menu.findItem(R.id.action_search)
        searchView = searchItem.actionView as SearchView
        searchView!!.setOnCloseListener(this)
        searchData()
        return true
    }

    override fun onClose(): Boolean {
        readData()
        return false
    }

    private fun hideHomeButton() {
        if (supportActionBar != null) {
            this@MainActivity.supportActionBar!!.setHomeButtonEnabled(false) // disable the button
            this@MainActivity.supportActionBar!!.setDisplayHomeAsUpEnabled(false) // remove the left caret
            this@MainActivity.supportActionBar!!.setDisplayShowHomeEnabled(false)
        }
    }

    companion object {
        const val ALPHA_FULL = 1.0f
        fun getBitmapFromVectorDrawable(context: Context?, drawableId: Int): Bitmap {
            var drawable = ContextCompat.getDrawable(context!!, drawableId)
            if (Build.VERSION.SDK_INT < Build.VERSION_CODES.LOLLIPOP) {
                drawable = DrawableCompat.wrap(drawable!!).mutate()
            }
            val bitmap = Bitmap.createBitmap(drawable!!.intrinsicWidth,
                    drawable.intrinsicHeight, Bitmap.Config.ARGB_8888)
            val canvas = Canvas(bitmap)
            drawable.setBounds(0, 0, canvas.width, canvas.height)
            drawable.draw(canvas)
            return bitmap
        }
    }
}